### ISSUE TITLE
Migration path from 1.3.x to 1.4.x

### ISSUE THREAD

**Issue by centur** [dotnet/orleans#2921]
Hi, all.
I'm not sure if this question was raised before, search over issues and gitter can't find anything...
Is there a migration guide from 1.3 to 1.4 ? 

Upgrading nugets breaks solution and as I can see in change notes - there are many breaking changes. but it's not always clear how to fix them. Would be nice to collect some of the knowledge in this issue so everyone can refer to.

1. All grains must have exactly one constructor that will be resolved with DI.

Previously grain is allowed to have empty ctor and unit-testing ctor e.g.:
```
...
public SubscriptionUpdaterGrain() { }
public SubscriptionUpdaterGrain(IGrainIdentity identity, IGrainRuntime runtime, SubscriptionUpdaterGrainState state, IStorage storage, IServiceLocator serviceLocator):
  base(identity, runtime, state, storage, serviceLocator) { }
...
```
in 1.4 Unit testing ctor must be removed and all grains must be converted to fully DI grains.
This is a big hit for us, we have plenty of grains and to upgrade - we need to convert them all including all unit tests where we were using DI ctor a lot...
As well as we need to register everything in Orleans DI. We are using Autofac, probably there is something as easy as `orleansContainer.RegisterAllFromAutofac(autofacContainer)` but I haven't studied this yet.

2. `TestClusterOptions` on `OrleansTestingHost` lost some properties (particularly `TestClusterOptions.DefaultTraceToConsole`) 

Proper migration path implies switching to a `TestCluster` , right ? In our case we can remove the testing host completely I think, as we are not running things in-silo anymore, but in general - curious to understand what other consequences may

3. Previously Registered serializers (e.g. for Json.NET objects) with 
`SerializationManager.SerializeInner(str, stream, typeof(string));` calls are now requiring `SerializationContext` instead of passing down `BinaryTokenStreamWriter stream` as it was before.

Migration: looks like they can be changed to :
`SerializationManager.SerializeInner(obj, new SerializationContext {StreamWriter = stream}, typeof(string));`

but previous `(string)SerializationManager.DeserializeInner(typeof(string), stream);` has deserialization context which is a bit different  `(string)SerializationManager.DeserializeInner(typeof(string), new DeserializationContext() {StreamReader = stream,CurrentObjectOffset = ???});`
Can anyone hint on whether CurrentObjectOffset would have a correct default value or it has to be initialized with something ?

NB: @sergeybykov on the benefits of semver from #2887 - nuget tools are based on semver compatibility so they allow users to upgrade from 1.3 to 1.4 expecting that nuget authors are doing semver too. So "safe to perform upgrade" from dev perspective becomes "everything is red" after some upgrades. There is a certain value to have early "breaking upgrade path" warnings.

**Comment by jdom** [comment]
Hi @centur, yeah, there are several breaking changes depending on how you use Orleans.
1) Pervasive DI: Since Orleans now always uses DI for constructing grains, and follows how AspNet Core works (where you have to have only 1 public constructor in your controller class), users should follow the same. The key here is that in needs 1 *public* constructor, but you could have other *internal* constructors you use from a testing assembly referenced by an InternalsVisiblesTo attribute.
Having said that, I'm in favor of being able to configure different grain creators in case this convention is not good for you (AspNet allows that too). Feel free to open an issue if the workaround doesn't sound good for you.
Not sure I understood the second part of the question. You do not need to register the grains themselves in the container (nor the registrations would be honored if you do), just the constructor dependencies need to be registered. If you were already using the default constructor and then using AutoFac after the fact to inject dependencies, you can continue to do so, as long as you have the default constructor be your only public constructor. Alternatively you can use the Startup configuration approach to replace the default container with AutoFac.

2) DefaultTraceToConsole was replaced by using the Microsoft.Extensions.Configuration package so that this value could come from different places, and not necessarily just a static property that applies to all. Nevertheless you can continue to do the same. See how we do it ourselves (on hindsight, we could simplify this, since it's a lot more flexible, but also complex): https://github.com/dotnet/orleans/blob/e0e6fe544bdfa1cc36c640f2de05a3fb9c601333/test/TestExtensions/TestDefaultConfiguration.cs#L36

3) @ReubenBond can probably answer this question, since he's being doing changes to be able to make SerializationManager (and the entire Orleans client) non-static. This will be part of the 1.5 release, but I guess some breaking changes made it into 1.4?

Let me know if you bumped into other issues. We can probably add some more info to the release notes

**Comment by kogir** [comment]
Regarding Json.Net, this is deployed and working well for me:

```C#
namespace RealArtists.ShipHub.ActorInterfaces.Serialization {
  using System;
  using System.Diagnostics.CodeAnalysis;
  using Newtonsoft.Json;
  using Newtonsoft.Json.Linq;
  using Orleans.CodeGeneration;
  using Orleans.Serialization;

  [Serializer(typeof(JToken))]
  // I guess maybe I have to register all descendant types too o_O
  [Serializer(typeof(JValue))]
  [Serializer(typeof(JRaw))]
  [Serializer(typeof(JContainer))]
  [Serializer(typeof(JArray))]
  [Serializer(typeof(JConstructor))]
  [Serializer(typeof(JObject))]
  [Serializer(typeof(JProperty))]
  internal class OrleansJTokenSerializer {
    // Can't make the class static or Orleans won't find it.
    private OrleansJTokenSerializer() { }

    private static readonly JsonSerializerSettings _JsonSerializerSettings = new JsonSerializerSettings {
      // The default value, DateParseHandling.DateTime, drops time zone information from DateTimeOffets.
      // This value appears to work well with both DateTimes (without time zone information) and DateTimeOffsets.
      DateParseHandling = DateParseHandling.DateTimeOffset,
      NullValueHandling = NullValueHandling.Ignore,
      Formatting = Formatting.Indented
    };

    [CopierMethod]
    [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "context")]
    [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
    public static object DeepCopier(object original, ICopyContext context) {
      // Even though JTokens *should* only be read, they can be edited. Let's play it safe.
      return ((JToken)original)?.DeepClone();
    }

    [SerializerMethod]
    [SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "expected")]
    [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
    public static void Serializer(object untypedInput, ISerializationContext context, Type expected) {
      var input = (JToken)untypedInput;
      var json = JsonConvert.SerializeObject(input, _JsonSerializerSettings);
      SerializationManager.SerializeInner(json, context, typeof(string));
    }

    [DeserializerMethod]
    [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
    public static object Deserializer(Type expected, IDeserializationContext context) {
      var json = (string)SerializationManager.DeserializeInner(typeof(string), context);
      return JsonConvert.DeserializeObject(json, expected, _JsonSerializerSettings);
    }
  }
}
```

**Comment by ReubenBond** [comment]
@kogir @centur Another option is to create an `IExternalSerializer` for it.

Something like this:
```C#
public class JsonObjectSerializer : IExternalSerializer
{
    private readonly JsonSerializerSettings settings;

    public JsonObjectSerializer(SerializationManager serializationManager, IGrainFactory grainFactory)
    {
        this.settings = OrleansJsonSerializer.GetDefaultSerializerSettings(serializationManager, grainFactory);
    }
        
    public void Initialize(Logger logger)
    {
    }
        
    public bool IsSupportedType(Type itemType)
    {
        return itemType == typeof(JToken) ||
               itemType == typeof(JValue) ||
               itemType == typeof(JRaw) ||
               itemType == typeof(JContainer) ||
               itemType == typeof(JArray) ||
               itemType == typeof(JConstructor) ||
               itemType == typeof(JObject) ||
               itemType == typeof(JProperty);
    }
        
    public object DeepCopy(object source, ICopyContext context)
    {
        return ((JToken)source)?.DeepClone();
    }
        
    public object Deserialize(Type expectedType, IDeserializationContext context)
    {
        var json = (string)SerializationManager.DeserializeInner(typeof(string), context);
        return JsonConvert.DeserializeObject(json, expectedType, this.settings);
    }
        
    public void Serialize(object item, ISerializationContext context, Type expectedType)
    {
        var input = (JToken)item;
        var json = JsonConvert.SerializeObject(input, this.settings);
        SerializationManager.SerializeInner(json, context, typeof(string));
    }
}
```

**Comment by kogir** [comment]
@ReubenBond Oh, nice! I'll investigate switching to that.

**Comment by centur** [comment]
Thanks, @jdom 
1. Got that, we'll try this and see how it aligns with the current test suit. Initially I tried to leave only unit test ctor (and remove parameterless default one) but it didn't worked out - not all of these parameters are registered out of the box in DI: `IGrainIdentity identity, IGrainRuntime runtime, T state, IStorage storage`. IIRC - it failed on `IGrainRuntime`. 
Which also raises the question - the grain mock up for unit testing is still the same - calling to the base ctor with 4 injectables as per above snippet ? 
The second part was more me thinking out loudly - we haven't used that abstracted DI yet, mainly using autofac. I think I have few snippets on how to configure MS Container (not sure whether they are still relevant or not), but I want to avoid copy-paste configuration of our AutoFac setup. So if there is a helper extension, that allows developers to configure Orleans DI container using existing popular DI containers in one line - it worth to mention here. 

2. 👍 looks nice, may be worth to switch our own configuration set up to this package 😄 
3. Yes, @ReubenBond pinged me on twitter and proposed some solutions too. Hat tip to him.

**Comment by jdom** [comment]
I'm not entirely clear on what you are saying. Are you currently using AutoFac with 1.3 or you are just planning to use it in the future? If you are already using it, are you actually injecting dependencies on your grains via constructor injection or not?

Regarding how to use AutoFac, it's similar to how it's done in ASP.NET, so this might be helpful: [Replacing the default services container](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection#replacing-the-default-services-container)

**Comment by centur** [comment]
We are using autofac to set up DI whereever it's possible, but in grains we started when there were no DI so  grains are using Autofac-based Service locator ( It's more a legacy than a practice we would consider "good") So we have our deps isolated in ctors with `DepA = ServiceLocator.Resolve<DepA>`, but they are not injected. We will inject them through ctor when will do the upgrade. And will use Autofac container as the only one container, so no re-registration needed. Your link provides a simple way to replace the entire container implementation instead of re-registering services from one container to another. 👍

**Comment by jdom** [comment]
Indeed, that's a better approach, to replace and use 1 container since it will respect lifetime of objects correctly (ie: if you register a singleton in both containers, you'll have 2 singletons if you ever resolve them from the 2 :smile:)

**Comment by centur** [comment]
I know, I wasn't aware that MS.Abstractions.DI has full support for this mode

**Comment by centur** [comment]
@ReubenBond your proposed serialization fix seems to be based on some custom version as in 1.4.1 SerializationManager is still static so it can't be used for injection in instance constructor. This is at least what VS tells me. But for our case we keep Json serialization settings as a copy on our own object so this is not a problem to get a ref from other object.

**Comment by ReubenBond** [comment]
@centur you're right, my code is based on changes introduced after 1.4.1

**Comment by centur** [comment]
Another question about serialization - 
I just found in docs, that orleans now has a built-in custom serializer for Json.NET objects:

[Advanced-Concepts: Serialization](http://dotnet.github.io/orleans/Documentation/Advanced-Concepts/Serialization.html)

`Orleans.Serialization.OrleansJsonSerializer`  which seems available in 1.4.1
So technically I don't need to register any custom stuff - json objects over the wire should work out of the box. Is this correct or it's already 1.5 beta doc ?

**Comment by ReubenBond** [comment]
@centur, that serializer uses JSON.NET to serialize objects, but when an external serializer is in use, that serializer is used for objects of all types which it claims to support via the `IsSupportedType(type)` method. Since `OrleansJsonSerializer` claims to support all types, it will be used preferentially for everything. If that's what you want, then go ahead, but if you solely want to serialize JSON.NET's own types (`JObject`, etc), then you'll need to add a custom serializer
